consensus/safety-rules/src/tests/suite.rs

```diff
@@ -600,3 +607,151 @@ fn test_sign_proposal_with_early_preferred_round(func: Callback) {
+    safety_rules.update(a1.block().quorum_cert()).unwrap();
+}
+
+// Tests for fetching a missing validator key from persistent storage.
+// Ideally under this circumstance, safety rules should panic. However,
+// at current stage of implementation the upper layer (i.e. RoundManager) is
+// unable to handle such failure (actually it is unable ANY failure
+// during key reconciliation!). For more fail cases, refer to debug messages in
+// reconcile_key()
+#[allow(dead_code)]
+fn test_a_missing_validator_key(func: Callback) {
+    let (mut safety_rules, signer) = func();
+    let (proof, genesis_qc) = make_genesis(&signer);
+    let round = genesis_qc.certified_block().round();
+
+    safety_rules.initialize(&proof).unwrap();
+
+    let a1 = test_utils::make_proposal_with_qc(round + 1, genesis_qc, &signer);
+
+    let mut next_epoch_state = EpochState::empty();
+    next_epoch_state.epoch = 1;
+    let rand_signer = ValidatorSigner::random([0xfu8; 32]);
+    next_epoch_state.verifier =
+        ValidatorVerifier::new_single(signer.author(), rand_signer.public_key());
+    let a2 = test_utils::make_proposal_with_parent_and_overrides(
+        vec![],
+        round + 2,
+        &a1,
+        Some(&a1),
+        &signer,
+        Some(1),
+        Some(next_epoch_state),
+    );
+    safety_rules.update(a2.block().quorum_cert()).unwrap_err();
+}
+
+fn test_a_missing_validator(func: Callback) {
+    // Testing for an validator missing from the validator set
+    // It does so by updating the safey rule to an epoch state, which does not contain the
+    // current validator. The validator later fails to verify QC since it is no longer in the
+    // validator set.
+
+    let (mut safety_rules, signer) = func();
+
+    let (proof, genesis_qc) = make_genesis(&signer);
+    let round = genesis_qc.certified_block().round();
+
+    safety_rules.initialize(&proof).unwrap();
+
+    let a1 = test_utils::make_proposal_with_qc(round + 1, genesis_qc, &signer);
+
+    let mut next_epoch_state = EpochState::empty();
+    next_epoch_state.epoch = 1;
+    let rand_signer = ValidatorSigner::random([0xfu8; 32]);
+    next_epoch_state.verifier =
+        ValidatorVerifier::new_single(rand_signer.author(), rand_signer.public_key());
+    let a2 = test_utils::make_proposal_with_parent_and_overrides(
+        vec![],
+        round + 2,
+        &a1,
+        Some(&a1),
+        &signer,
+        Some(1),
+        Some(next_epoch_state),
+    );
+    safety_rules.update(a2.block().quorum_cert()).unwrap();
+
+    let a3 = test_utils::make_proposal_with_parent_and_overrides(
+        vec![],
+        round + 3,
+        &a2,
+        Some(&a2),
+        &signer,
+        Some(1),
+        None,
+    );
+    let err = safety_rules.update(a3.block().quorum_cert()).unwrap_err();
+    assert_eq!(
+        err,
+        Error::InvalidQuorumCertificate("Fail to verify QuorumCert".into())
+    );
+}
+
+fn test_reconcile_key(_func: Callback) {
```

Comment:
If you follow the feedback given in this and previous reviews, the recommendation has been that there is "None" set for the current validator signer and hence at initialization we will already be testing reconciliation. I would argue we want just want one positive test that shows we can retrieve the key even if a rotation has happened in storage. And then we rotate to that key in a QC and reconciliation updates to the new key. Thoughts?

Reply:
Sorry about the resolving conversation thing, which costs you more time to realize the changes -- I wasn't aware of the right practice and will do it correctly afterwards. 
I agree with the test method. 
To be more specific, the TODO items are: 
1. refactor the constructor, 
2. reduce test cases into 2 (1 pos + 1 neg).
On them..

Reply:
I think the tests are okay, because they test distinct things. I do wonder the reconcile is a little long, but I'm okay with just wrapping up at this point :)